task/buildah-rhtap/0.1/buildah-rhtap.yaml

```diff
@@ -0,0 +1,237 @@
+        dockerfile_path=$(mktemp --suffix=-Dockerfile)
+        http_code=$(curl -s -L -w "%{http_code}" --output "$dockerfile_path" "$DOCKERFILE")
+        if [ $http_code != 200 ]; then
+          echo "No Dockerfile is fetched. Server responds $http_code"
+          exit 1
+        fi
+        http_code=$(curl -s -L -w "%{http_code}" --output "$dockerfile_path.dockerignore.tmp" "$DOCKERFILE.dockerignore")
+        if [ $http_code = 200 ]; then
+          echo "Fetched .dockerignore from $DOCKERFILE.dockerignore"
+          mv "$dockerfile_path.dockerignore.tmp" $SOURCE_CODE_DIR/$CONTEXT/.dockerignore
+        fi
+      else
+        echo "Cannot find Dockerfile $DOCKERFILE"
+        exit 1
+      fi
+
+      # Fixing group permission on /var/lib/containers
+      chown root:root /var/lib/containers
+
+      sed -i 's/^\s*short-name-mode\s*=\s*.*/short-name-mode = "disabled"/' /etc/containers/registries.conf
+
+      LABELS=(
+        "--label" "build-date=$(date -u +'%Y-%m-%dT%H:%M:%S')"
+        "--label" "architecture=$(uname -m)"
+        "--label" "vcs-type=git"
+      )
+      [ -n "$COMMIT_SHA" ] && LABELS+=("--label" "vcs-ref=$COMMIT_SHA")
+      [ -n "$IMAGE_EXPIRES_AFTER" ] && LABELS+=("--label" "quay.expires-after=$IMAGE_EXPIRES_AFTER")
+
+      buildah build \
+        ${LABELS[@]} \
+        --tls-verify=$TLSVERIFY --no-cache \
+        --ulimit nofile=4096:4096 \
+        -f "$dockerfile_path" -t $IMAGE $SOURCE_CODE_DIR/$CONTEXT
+
+      container=$(buildah from --pull-never $IMAGE)
+      buildah mount $container | tee /workspace/container_path
+      echo $container > /workspace/container_name
+    securityContext:
+      capabilities:
+        add:
+          - SETFCAP
+    volumeMounts:
+    - mountPath: /var/lib/containers
+      name: varlibcontainers
+    workingDir: $(workspaces.source.path)
+
+  - name: sbom-syft-generate
+    image: quay.io/redhat-appstudio/syft:v0.98.0@sha256:4d3856e6a2622700b9a9d5d74d9aaf5d8a55671653f80bf6c636677658680ede
+    script: |
+      syft dir:$(workspaces.source.path)/source --output cyclonedx-json=$(workspaces.source.path)/sbom-source.json
+      find $(cat /workspace/container_path) -xtype l -delete
+      syft dir:$(cat /workspace/container_path) --output cyclonedx-json=$(workspaces.source.path)/sbom-image.json
+    volumeMounts:
+    - mountPath: /var/lib/containers
+      name: varlibcontainers
+
+  - name: merge-syft-sboms
+    image: registry.access.redhat.com/ubi9/python-39:1-158@sha256:967000729b17efdea309e297f4b1961c38b902a1ef18f6d886b8086c2a12f01f
+    script: |
+      #!/bin/python3
+      import json
+
+      # load SBOMs
+      with open("./sbom-image.json") as f:
+        image_sbom = json.load(f)
+
+      with open("./sbom-source.json") as f:
+        source_sbom = json.load(f)
+
+      # fetch unique components from available SBOMs
+      def get_identifier(component):
+        return component["name"] + '@' + component.get("version", "")
```

Comment:
I changed the algorithm so it looks at the `purl` now. Thing is, there is one case in which the purl is always absent: operating-system components. That should be fine, as they're only reported once in the image scan, but it led me to allow any component that lacks a purl to be present in the final SBOM (at risk of duplication).

Reply:
I don't think this inline Python script looks good, though, so I think we should improve Cachi2's merge [script](https://github.com/containerbuildsystem/cachi2/blob/main/utils/merge_syft_sbom.py) to be more generalistic, able to handle any CycloneDX SBOM regardless of the generation tool.

Reply:
Trying to do general SBOM de-duplication will be rough, because purls => packages is an N:1 relation, what with all the optional qualifiers etc. Some duplicates will always remain.

And if the SBOM generation tool is not careful or doesn't care, purls => packages can also be 1:M. For example, you can't be sure if `pkg:golang/sigs.k8s.io/cluster-api-provider-aws/v2` really means the original module or the [`github.com/openshift/cluster-api-provider-aws`](https://github.com/openshift/cluster-api-provider-aws/blob/4c2d9b8e1626337cdbe8cdd5a0ad43c2d520ecd1/go.mod#L1C8-L1C47) fork (with cachi2 you probably can, with Syft you can't). You also can't be sure if the tool neglected to include a `?repository` qualifier for a package that comes from a non-default repository, `?vcs_url` for a package that comes straight from git, etc. In this case, by trusting the purl you end up discarding non-duplicate packages, which may or may not be fine.

Reply:
Yeah, excelent point. So I think we'll need to keep doing a tool-based dedup, since we can better predict its behavior then. Maybe we could move this script to the rh-syft repo in the future? I don't think it belongs in Cachi2.

Reply:
> Maybe we could move this script to the rh-syft repo in the future?

The fewer things we productize (and have to support) the better. I wanted the `syft` executable to be the only thing in the image, but I may end up including a ubi-micro base. But probably nothing past that.

If we do need a script long-term, it should probably be tightly coupled to the task, versioned together with the task etc. Though it shouldn't be inlined in the yaml and it should be tested.

Reply:
Btw, one of the cataloggers that `syft` includes is `sbom`. Maybe we don't need a script for syft X syft deduplication at all

Reply:
I did some early testing and the sobm cataloger seems promising. I will file a story to investigate it further.